### PR TITLE
🪄 Steps

### PR DESCRIPTION
**Pull request by jpwilliams**
# ✨ Step functions

Adds initial step function support to the SDK, allowing users to create complex flows with some basic tooling.

```ts
inngest.createStepFunction(
  "A Step Function",
  "app/user.created",
  ({ event, tools }) => {
    const seenEvent = tools.waitForEvent("app/user.seen", {
      match: ["data.email"],
      timeout: "2d",
    });

    if (!seenEvent) {
      tools.run("Something", () => {
        /* ... */
      });
    }

    tools.sleep("2m");

    tools.run("Async task", async () => {
      /* ... */
    });
  }
);
```

We have three tools available in the MVP:

- `waitForEvent` to pause execution until an event is seen or until the timeout is met
- `run` to run business logic
- `sleep` to pause execution for a set amount of time
- `sleepUntil` to pause execution until a specific date (just figures out the TTL)

Some notable caveats:

- `waitForEvent`'s `if` is difficult to type and therefore not ideal for the user; they likely need docs to understand how to do anything complex. @djfarrelly's array idea helped a lot with covering common approaches.

## Working on

- [x] Tests to ensure responses from step functions are correct and reproducible
- [x] Individual tool unit tests framework

### CODE REVIEW

File: src/components/InngestStepTools.ts [hunk 1]
```diff
@@ -0,0 +1,418 @@
...
+export const createStepTools = <
+  Events extends Record<string, EventPayload>,
+  TriggeringEvent extends keyof Events
+>(
+  _opStack: OpStack
+) => {
+  /**
+   * Controls whether these toolsets are active or not. Exposed tools should
+   * watch this value to decide what action to perform. It is most likely that
+   * if this boolean is `false` then they should immediately interrupt flow.
```
Comment (tonyhb) [inline src/components/InngestStepTools.ts:36]: When/why is this necessary?

File: src/components/InngestStepTools.ts [hunk 2]
```diff
@@ -0,0 +1,418 @@
...
+        }
+
+        if (opts?.match) {
+          if (typeof opts.match === "string") {
+            matchOpts.match = `event.${opts.match} == async.${opts.match}`;
+          } else {
+            matchOpts.match = `async.${opts.match[0]} == ${
+              typeof opts.match[1] === "string"
+                ? `'${opts.match[1]}'`
+                : // eslint-disable-next-line @typescript-eslint/restrict-template-expressions
```
Comment (tonyhb) [inline src/components/InngestStepTools.ts:247]: `expression((event, matching) => `${event("data.name")} == ${matching("foo.bar")}`);`

**Review by tonyhb — COMMENTED** [review]
Amazing work!!1 Holy crap this is exciting.

- Let's read from `payload.steps` instead of the stack for any particular tool!
- Can we remove `match: [a, b]` overall, and only have the string shortcut until we figure out a nice API to write CEL expressions?

Some questions:

- The presence of a sleep in the stack indicates that the sleep/wait is done, right?  If so, can this be annotated as a comment?
- Why do we have submitOp and setPendingOp?  Can we merge?
- Can we rename `active` to `memoized` or `readingFromStack`?


This work, including the typing, is gold.  Incredible!

**Review by djfarrelly — APPROVED** [review]
This is an awesome PR. The work with the tools and the op stack is fantastic! Also the documentation is really good and made it easy to review this. I left a few comments that aren't blockers.

I think I was originally going to ask the question if `tools` should be a second arg instead of being in the first arg object with `event`, but I've come around to this and it seems simpler to have them all in a single destructured object arg. It's going to be more intuitive the way that you wrote it and I'll update the beta landing page on the site to match!

File: package.json [hunk 3]
```diff
@@ -19,7 +19,7 @@
     "pb:version": "genversion --semi --double --es6 ./src/version.ts",
     "pb:landing": "yarn build:landing && node -e 'const page = JSON.stringify(require(\"fs\").readFileSync(\"./landing/dist/index.html\").toString()); console.log(\"export const landing = \" + page);' > ./src/landing.ts && npx prettier ./src/landing.ts --write",
     "build": "yarn run clean && tsc --project tsconfig.build.json",
-    "test": "node --expose-gc --max-old-space-size=4096 ./node_modules/.bin/jest --silent --logHeapUsage --maxWorkers=8",
+    "test": "node --expose-gc --max-old-space-size=4096 ./node_modules/.bin/jest --silent --logHeapUsage --maxWorkers=8 --coverage",
```
[the remark this answers is not in the document]
Comment (jpwilliams) [reply package.json:22]: Told you I'd sneak it in! 😄

image

File: src/components/InngestFunction.ts [hunk 4]
```diff
@@ -72,46 +77,136 @@ export class InngestFunction<Events extends Record<string, EventPayload>> {
...
-              id: stepId,
-              name: stepId,
-              runtime: {
-                type: "http",
-                url: url.href,
-              },
-            },
-          };
+      steps: {
+        step: {
```
[the remark this answers is not in the document]
Comment (jpwilliams) [reply src/components/InngestFunction.ts:91]: Yes! Missed this one - thank you :)

File: src/components/InngestStepTools.ts [hunk 5]
```diff
@@ -0,0 +1,473 @@
...
+  OpStack,
+  StepOpCode,
+  TimeStr,
+} from "../types";
+
+/**
+ * A unique class used to interrupt the flow of a step. It is intended to be
+ * thrown and caught using `instanceof StepFlowInterrupt`.
+ */
+export class StepFlowInterrupt {}
```
[the remark this answers is not in the document]
Comment (jpwilliams) [reply src/components/InngestStepTools.ts:18]: Yeah doesn't really matter - we don't want to ever log it or use it other than to do `err instanceof StepFlowInterrupt`; no need to extend `Error`, provide a message, etc.

File: src/components/InngestStepTools.test.ts [hunk 6]
```diff
@@ -0,0 +1,197 @@
...
+      opts: {
+        match: "name == 123",
+      },
+      hash: "9d3fe213c82f60f52e91b036388fa7851b7ef6b3",
+    });
+  });
+
+  test("prioritise `match` statement if both `match` and `if` given", async () => {
+    expect(() =>
+      waitForEvent("event", { match: "name", if: "name == 123" })
```
[the remark this answers is not in the document]
Comment (jpwilliams) [reply src/components/InngestStepTools.test.ts:93]: @djfarrelly Not sure what you mean? Do you mean if the opts for `waitForEvent` are incorrect?

File: src/components/InngestStepTools.ts [hunk 7]
```diff
@@ -0,0 +1,473 @@
...
+         */
+        opts
+      ) => {
+        const matchOpts: { ttl?: string; match?: string } = {};
+
+        if (opts?.timeout) {
+          matchOpts.ttl =
+            typeof opts.timeout === "string"
+              ? opts.timeout
+              : dateToTimeStr(opts.timeout);
```
[the remark this answers is not in the document]
Comment (jpwilliams) [reply src/components/InngestStepTools.ts:269]: @djfarrelly Not sure what you mean by "_this timeout could float_," but this function will run, compare the date given to now (`new Date()`), create a time string from it (e.g. `"1h15m10s"`), then send that time string to Inngest and wait to be called again.

There's no local timeout that happens if that's what you were getting at? Sleepy Saturday morning here so forgive me if I've misunderstood. 😄

File: src/components/InngestStepTools.test.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (djfarrelly) [reply src/components/InngestStepTools.test.ts:93]: Yeah @jpwilliams, exactly. If we're ignoring one of the options, should we throw and error that says something like "Can only pass match or if, but not both, defaulting to match?" I'm just thinking how to share feedback to the user since they may not realize in this case that their `if` is being superseded by the `match`

File: src/components/InngestStepTools.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (djfarrelly) [reply src/components/InngestStepTools.ts:269]: Sorry, my thought wasn't very clear. I was trying to think about ways that this could produce unintended results, but I think it's not something we can handle in code here, the dev just needs to understand how it works.

The idea in my head is that we can accept a `Date`, which feels exact, but then we convert to something that is relative (ttl), so I was trying to think in what cases (retries, etc.) where a user providing a Date could expect something to be more exact (e.g. scheduling) than it will real be. I could see if a user decides to use a timestamp from the event payload to pass to the TTL, retries may fail depending on the data, but that may just be something devs need to learn to handle in code, not something we could necessarily optimize for in the library. We should definitely add pointers to out docs for using this tool :)

As a side note: I think in our examples, we should use [`dayjs`](https://day.js.org/) to show this could be used.

File: src/helpers/strings.ts [hunk 8]
```diff
@@ -11,3 +11,75 @@ export const slugify = (str: string): string => {
...
+  }
+
+  /**
+   * TODO In this eventuality, should we smartly skip the sleep?
+   *
+   * We'd have to return two ops to Inngest, this being skipped and the
+   * next.
+   */
+  if (date <= now) {
+    throw new Error("Cannot sleep until a time in the past");
```
Comment (djfarrelly) [inline src/helpers/strings.ts:66]: @jpwilliams, I am assuming this error will appear in the logs for a given function run in the Inngest dashboard. Could we include the ISO8601 timestamp in the error message? That may help a user determine what bad value in the past that their code may have produced.

Comment (tonyhb) [reply src/helpers/strings.ts:66]: Or, if this is a timing/race condition, should we warn and then continue with no sleep?

File: src/components/InngestStepTools.test.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (jpwilliams) [reply src/components/InngestStepTools.test.ts:93]: I'll make the opts so it has to be either one or the other, but not both.

File: src/components/InngestStepTools.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (jpwilliams) [reply src/components/InngestStepTools.ts:269]: Gotchya. Aye, I think you, @djfarrelly, and @tonyhb have converged on this idea! I had been matching the TTL field in the executor for the moment, but dates everywhere would be much nicer, even under the hood.

If we move to pass full timestamps to the executor after this, we don't have to worry so much about confusing conversions. We still have clock skew problems technically, but aye there's not too much we can do to remedy that other than potentially mentioning somewhere that any relative times passed are relative _to the machine running the code_.

File: src/helpers/strings.ts (hunk 8 above)
Comment (jpwilliams) [reply src/helpers/strings.ts:66]: Mm I'm not a fan of this `throw` at all!

The executor should probably hold the knowledge here, so perhaps we always push the operation to the executor regardless of when we think the date is, then the executor can call right back if it needs to.

File: src/components/InngestStepTools.test.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (jpwilliams) [reply src/components/InngestStepTools.test.ts:93]: image
image
image
